pep-9999.rst

```diff
@@ -0,0 +1,1176 @@
+interpreter and prevents breaking system software.
+
+.. _virtualenv: https://virtualenv.pypa.io/en/latest/
+
+In some cases, however, it's useful and intentional to
+install a Python package from outside of the distribution
+that influences the behavior of distribution-shipped
+commands. This is common in the case of software like Sphinx
+or Ansible which have a mechanism for writing
+Python-language extensions. A user may want to use their
+distribution's version of the base software (for reasons of
+paid support or security updates) but install a small
+extension from PyPI, and they'd want that extension to be
+importable by the software in their base system.
+
+While this continues to carry the risk of installing a newer
+version of a dependency than the operating system expects or
+otherwise negatively affecting the behavior of an
+application, it does not need to carry the risk of removing
+files from the operating system. A tool like pip should be
+able to install packages in some directory on the default
+``sys.path``, if specifically requested, without deleting
+files owned by the system's package manager.
+
+Therefore, this PEP proposes two things.
+
+First, it proposes **a way for distributors of a Python
+interpreter to mark that interpreter as having its packages
+managed by means external to Python**, such that
+Python-specific tools like pip should not change the
+installed packages in the interpreter's global ``sys.path``
+in any way (add, upgrade/downgrade, or remove) unless
+specifically overridden.  It also provides a means for the
+distributor to indicate how to use a virtual environment as
+an alternative.
+
+This is an opt-in mechanism: by default, the Python
+interpreter compiled from upstream sources will not be so
+marked, and so running ``pip install`` with a self-compiled
+interpreter, or with a distribution that has not explicitly
+marked its interpreter, will work as it always has worked.
+
+Second, it sets the rule that when installing packages to an
+interpreter's global context (either to an unmarked
+interpreter, or if overriding the marking),
+**Python-specific package managers should modify or delete
+files only within the directories of the sysconfig
+scheme in which they would create files**. This permits a
+distributor of a Python interpreter to set up two
+directories, one for its own managed packages, and one for
+unmanaged packages installed by the end user, and ensure
+that installing unmanaged packages will not delete (or
+overwrite) files owned by the external package manager.
+
+Rationale
+=========
+
+As described in detail in the next section, the first
+behavior change involves creating a marker file named
+``EXTERNALLY-MANAGED``, whose presence indicates that
+non-virtual-environment package installations are managed by
+some means external to Python, such as a distribution's
+package manager. This file is specified
+to live in the ``stdlib`` directory in the default
```

Comment:
Why not `platlib` or `purelib`, which will automatically handle the venv case?

Reply:
Because we still want to block `--user` installs in a virtual environment. pip currently blocks `--user` in a virtual environment, but other installers may not implement the same. Another consideration is that files in `platlib` and `purelib` could be overriden by a package, which is problematic. Finally, this makes scheme-switching more complicated because each scheme has different `platlib` and `purelib` values (should a tool respect the file in the *target* scheme’s `purelib` or the default scheme’s?) while the `stdlib` is always the same.

Reply:
I didn't see anything here that would prevent you from keeping the venv case anyway, or any reason to make that universal (what if you pipx run a tool for managing packages in other environments?).

Presumably even in `stdlib` a package could bring its own copy of the file and start enforcing restrictions. That should probably just be a "don't do that" scenario, since I can actually think of valid reasons to do it (not on PyPI, of course).

I thought it was explicit that you always rely on the default scheme? But I also don't see why you wouldn't look at the directory that you're about to install into and decide then. That's going to make the most sense for everyone involved. (Merging `purelib` and `platlib` I assume is more complex, but then I have to assume that installers currently have a way to handle this - it's not an issue on Windows, so I've never actually had to deeply understand the practical difference).

Reply:
> Presumably even in `stdlib` a package could bring its own copy of the file and start enforcing restrictions. That should probably just be a "don't do that" scenario, since I can actually think of valid reasons to do it (not on PyPI, of course).

I think that’s a different preference then. From the discussions I was involved in during the PyCon US sprint, it was stated as an explicit goal that other authors *don’t* want this to be possible (I don’t have strong preferences personally).

Reply:
Then either installers have to filter the files they're willing to extract (which I think is fine), or the setting gets moved out of the install directory and into some other configuration location (which I also think is fine). Otherwise there's no way currently to stop a rogue package from adding the file (though as I say, I think that's fine too, and we just file bugs and aggressively fork anyone who tries to do it on PyPI :-) ). Nobody is _removing_ or replacing an existing file without unsupported hacking, of course.

It still seems more consistent with files like `.htaccess` and `.gitignore` to have this configuration affect the directory that it's in. The rejected idea for this presumes that the marker can be found in any `sys.path` entry and then apply to the installation as a whole, but it's not clear whether it was considered to have the marker be discovered (either anywhere, or only in the installers intended target directory) and specify either which schemes/paths may (or may not) be used.

Reply:
Because if you're installing to the user scheme (with e.g. `pip install --user`), then `platlib` and `purelib` point to your home directory, but we still want to prevent people from breaking system applications by installing things there (by default).